### DISCUSSION TITLE
Trying to add new ciphers and need help

### DISCUSSION THREAD

**Discussion by rgolangh** [ipxe/ipxe#729] [General]
I'm trying to add ciphers to the tls stack and I need some help.
Specifically I want to support tls 1.2 ECDHE with AES 128 GCM.  
I want to use openssl functions and implement the server key exchange, and the client key exchange
but I'm failing to use openssl or even libsodium. The failure is that 'make' is looping in the 'DEPS' stage:
1. add the include `#include <openssl/engine.h>` to `net/tls.c`
2. make sure openssl-devel is installed in the system
``` $ make bin/ipxe.iso
  ...
  [DEPS] net/tls.c
  [DEPS] net/tls.c
  [DEPS] net/tls.c
  [DEPS] net/tls.c
  [DEPS] net/tls.c
  [DEPS] net/tls.c
```
Make version: GNU Make 4.3
gcc version 12.1.1    (I cherry-picked a patch to suppress the warnings)

I tried adding `-d` to the make command, and it says that is pruning the makefiles and reexecuting the make. I have no idea what's
invoking that make over and over again. 
I do see that the file in `bin/deps/net/tls.c.d` is recreated over and over again.

I'll supply any info needed.

**Comment by NiKiZe** [comment]
I think you will have a hard time using external libraries. But there was a PR for adding QR code, maybe that could be used?

**Comment by mcb30** [comment]
iPXE is a freestanding binary and essentially cannot use any external libraries such as OpenSSL.

While it is possible to hack the build system to work around this (deliberate) limitation, the end result will probably not be what you need. iPXE code is extremely small with a very custom runtime environment, and you will find that any attempt to directly link in a general-purpose library such as OpenSSL would increase the binary size by a few thousand percent, and hit a variety of API incompatibility issues.

**Comment by rgolangh** [comment]
was this solved in https://github.com/ipxe/ipxe/commit/963ec1c4f379a49cf37d01472a770fff8e47470c#diff-6afc7c3ae67a546697b821c675883059a147c3840042e5c37ce8867349365f45R34 ?

**Comment by mcb30** [comment]
> was this solved in https://github.com/ipxe/ipxe/commit/963ec1c4f379a49cf37d01472a770fff8e47470c#diff-6afc7c3ae67a546697b821c675883059a147c3840042e5c37ce8867349365f45R34 ?

Yes, ECDHE key exchange is now supported in iPXE.
